File: apps/server-mac/PrismServer/Services/ConfigStore.swift
import Foundation

final class ConfigStore {
    private static let legacyDefaultAPIPort = 8787
    private static let legacyDefaultWebPort = 3000

    let applicationSupportDirectory: URL
    let logDirectory: URL

    private let envFileURL: URL

    init(fileManager: FileManager = .default) {
        let supportRoot = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        applicationSupportDirectory = supportRoot.appendingPathComponent("Prism", isDirectory: true)
        logDirectory = fileManager.urls(for: .libraryDirectory, in: .userDomainMask).first!
            .appendingPathComponent("Logs", isDirectory: true)
            .appendingPathComponent("Prism", isDirectory: true)
        envFileURL = applicationSupportDirectory.appendingPathComponent(".env")
    }

    func load() -> ServerConfig {
        guard
            let raw = try? String(contentsOf: envFileURL, encoding: .utf8)
        else {
            return .defaults
        }

        let env = Self.parseEnv(raw)
        var config = ServerConfig.defaults
        config.serverName = env["PRISM_SERVER_NAME"] ?? config.serverName
        config.apiPort = Self.readInt(env["API_PORT"], fallback: config.apiPort)
        config.webPort = Self.readInt(env["WEB_PORT"], fallback: config.webPort)
        if config.apiPort == Self.legacyDefaultAPIPort,
           config.webPort == Self.legacyDefaultWebPort {
            config.apiPort = ServerConfig.defaults.apiPort
            config.webPort = ServerConfig.defaults.webPort
        }
        config.discoveryEnabled = Self.readBool(env["PRISM_DISCOVERY_ENABLED"], fallback: config.discoveryEnabled)
        config.sessionCookieName = env["SESSION_COOKIE_NAME"] ?? config.sessionCookieName
        config.sessionTtlHours = Self.readInt(env["SESSION_TTL_HOURS"], fallback: config.sessionTtlHours)
        config.encryptionMasterKey = env["ENCRYPTION_MASTER_KEY"] ?? config.encryptionMasterKey
        config.ollamaHost = env["OLLAMA_HOST"] ?? config.ollamaHost
        config.ollamaModel = env["OLLAMA_MODEL"] ?? config.ollamaModel
        config.qdrantURL = env["QDRANT_URL"] ?? config.qdrantURL
        config.openAIAPIKey = env["OPENAI_API_KEY"] ?? config.openAIAPIKey
        return config
    }

    func save(_ config: ServerConfig) throws {
        try FileManager.default.createDirectory(at: applicationSupportDirectory, withIntermediateDirectories: true)
        try FileManager.default.createDirectory(at: logDirectory, withIntermediateDirectories: true)

        let contents = """
        API_PORT=\(config.apiPort)
        WEB_PORT=\(config.webPort)
        PRISM_SERVER_NAME=\(config.serverName)
        PRISM_DISCOVERY_ENABLED=\(config.discoveryEnabled ? "true" : "false")
        SESSION_COOKIE_NAME=\(config.sessionCookieName)
        SESSION_TTL_HOURS=\(config.sessionTtlHours)
        ENCRYPTION_MASTER_KEY=\(config.encryptionMasterKey)
        OLLAMA_HOST=\(config.ollamaHost)
        OLLAMA_MODEL=\(config.ollamaModel)
        QDRANT_URL=\(config.qdrantURL)
        OPENAI_API_KEY=\(config.openAIAPIKey)
        NEXT_TELEMETRY_DISABLED=1
        """

        try contents.write(to: envFileURL, atomically: true, encoding: .utf8)
    }

    static func parseEnv(_ raw: String) -> [String: String] {
        raw.split(whereSeparator: \.isNewline).reduce(into: [String: String]()) { result, line in
            let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !trimmed.isEmpty, !trimmed.hasPrefix("#") else { return }
            let parts = trimmed.split(separator: "=", maxSplits: 1, omittingEmptySubsequences: false)
            guard parts.count == 2 else { return }
            result[String(parts[0])] = String(parts[1])
        }
    }

    private static func readInt(_ value: String?, fallback: Int) -> Int {
        guard let value, let intValue = Int(value) else {
            return fallback
        }
        return intValue
    }

    private static func readBool(_ value: String?, fallback: Bool) -> Bool {
        guard let value else {
            return fallback
        }
        switch value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
        case "1", "true", "yes", "on":
            return true
        case "0", "false", "no", "off":
            return false
        default:
            return fallback
        }
    }
}

File: apps/server-windows/core/Services/ConfigStore.cs
using PrismServer.Core.Models;

namespace PrismServer.Core.Services;

public sealed class ConfigStore
{
    private const int LegacyDefaultApiPort = 8787;
    private const int LegacyDefaultWebPort = 3000;

    private readonly PrismPaths _paths;

    public ConfigStore(PrismPaths paths)
    {
        _paths = paths;
    }

    public string ApplicationSupportDirectory => _paths.ApplicationSupportDirectory;
    public string LogDirectory => _paths.LogDirectory;
    public string EnvFilePath => _paths.EnvFilePath;

    public ServerConfig Load()
    {
        if (!File.Exists(_paths.EnvFilePath))
        {
            return ServerConfig.Defaults;
        }

        var env = ParseEnv(File.ReadAllText(_paths.EnvFilePath));
        var defaults = ServerConfig.Defaults;
        var config = defaults with
        {
            ServerName = ReadString(env, "PRISM_SERVER_NAME", defaults.ServerName),
            ApiPort = ReadInt(env, "API_PORT", defaults.ApiPort),
            WebPort = ReadInt(env, "WEB_PORT", defaults.WebPort),
            DiscoveryEnabled = ReadBool(env, "PRISM_DISCOVERY_ENABLED", defaults.DiscoveryEnabled),
            SessionCookieName = ReadString(env, "SESSION_COOKIE_NAME", defaults.SessionCookieName),
            SessionTtlHours = ReadInt(env, "SESSION_TTL_HOURS", defaults.SessionTtlHours),
            EncryptionMasterKey = ReadString(env, "ENCRYPTION_MASTER_KEY", defaults.EncryptionMasterKey),
            OllamaHost = ReadString(env, "OLLAMA_HOST", defaults.OllamaHost),
            OllamaModel = ReadString(env, "OLLAMA_MODEL", defaults.OllamaModel),
            QdrantUrl = ReadString(env, "QDRANT_URL", defaults.QdrantUrl),
            OpenAiApiKey = ReadString(env, "OPENAI_API_KEY", defaults.OpenAiApiKey)
        };

        if (config.ApiPort == LegacyDefaultApiPort && config.WebPort == LegacyDefaultWebPort)
        {
            config = config with
            {
                ApiPort = defaults.ApiPort,
                WebPort = defaults.WebPort
            };
        }

        return config;
    }

    public void Save(ServerConfig config)
    {
        Directory.CreateDirectory(_paths.ApplicationSupportDirectory);
        Directory.CreateDirectory(_paths.LogDirectory);

        var lines = new[]
        {
            $"API_PORT={config.ApiPort}",
            $"WEB_PORT={config.WebPort}",
            $"PRISM_SERVER_NAME={config.ServerName}",
            $"PRISM_DISCOVERY_ENABLED={(config.DiscoveryEnabled ? "true" : "false")}",
            $"SESSION_COOKIE_NAME={config.SessionCookieName}",
            $"SESSION_TTL_HOURS={config.SessionTtlHours}",
            $"ENCRYPTION_MASTER_KEY={config.EncryptionMasterKey}",
            $"OLLAMA_HOST={config.OllamaHost}",
            $"OLLAMA_MODEL={config.OllamaModel}",
            $"QDRANT_URL={config.QdrantUrl}",
            $"OPENAI_API_KEY={config.OpenAiApiKey}",
            "NEXT_TELEMETRY_DISABLED=1"
        };

        File.WriteAllText(_paths.EnvFilePath, string.Join(Environment.NewLine, lines) + Environment.NewLine);
    }

    public static Dictionary<string, string> ParseEnv(string raw)
    {
        var env = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        using var reader = new StringReader(raw);
        string? line;
        while ((line = reader.ReadLine()) is not null)
        {
            var trimmed = line.Trim();
            if (trimmed.Length == 0 || trimmed.StartsWith('#'))
            {
                continue;
            }

            var index = trimmed.IndexOf('=');
            if (index <= 0)
            {
                continue;
            }

            env[trimmed[..index]] = trimmed[(index + 1)..];
        }

        return env;
    }

    private static string ReadString(Dictionary<string, string> env, string key, string fallback) =>
        env.TryGetValue(key, out var value) ? value : fallback;

    private static int ReadInt(Dictionary<string, string> env, string key, int fallback) =>
        env.TryGetValue(key, out var value) && int.TryParse(value, out var parsed) ? parsed : fallback;

    private static bool ReadBool(Dictionary<string, string> env, string key, bool fallback)
    {
        if (!env.TryGetValue(key, out var value))
        {
            return fallback;
        }

        return value.Trim().ToLowerInvariant() switch
        {
            "1" or "true" or "yes" or "on" => true,
            "0" or "false" or "no" or "off" => false,
            _ => fallback
        };
    }
}
